jax/_src/interpreters/pxla.py

```diff
@@ -3225,21 +3225,49 @@ def create_cpp_call(self, params: stages.CompiledCallParams):
       return None
 
     def aot_cache_miss(*args, **kwargs):
-      # args do not include the const args
+      # args do not include the const args.
+      # We fix the const args to account for 64-bit const args when the
+      # lowering and execution use different `enable_x64` mode.
       # See https://docs.jax.dev/en/latest/internals/constants.html.
-      outs, out_flat, args_flat = stages.Compiled.call(params, *args, **kwargs)
+      fixed_const_args = []
+      for c, c_arg_info, c_shard, c_layout in \
+          zip(params.const_args,
+              self._all_args_info.in_avals[:len(params.const_args)],
+              self._in_shardings[:len(params.const_args)],
+              self._dispatch_in_layouts[:len(params.const_args)]):
+
+        can_c = xla.canonicalize_dtype(c)
+        # can_c does not match the dtype for which the code was compiled
+        if can_c.dtype != c_arg_info.dtype:
```

Comment:
An alternative here would be to tag `Lowered` and `Compiled` with the value of the `enable_x64` flag. I chose to detect the mismatch because this results in fewer user-visible changes to data structures.